src/connectors/s3.rs

```diff
@@ -0,0 +1,373 @@
+            if let Ok(new_client) = self.client_receiver.recv().await {
+                self.client = Some(new_client);
+            }
+        } else if let Ok(new_client) = self.client_receiver.try_recv() {
+            self.client = Some(new_client);
+        }
+
+        let ingest_id = event.ingest_ns;
+
+        for (event, meta) in event.value_meta_iter() {
+            // Handle no key in meta.
+            let object_key = match meta.get("key").map(|value| value.to_string()) {
+                Some(key) => key,
+                None => {
+                    self.current_key.clear();
+                    error!("s3: Missing Key Meta Data in Event");
+                    return Ok(SinkReply::FAIL);
+                }
+            };
+
+            if object_key != self.current_key {
+                // Complete the uploads for previous key.
+                self.set_new_key(object_key).await?;
+            }
+
+            // Handle the aggregation.
+            for data in serializer.serialize(event, ingest_id)?.into_iter() {
+                self.buffer.extend(data);
+                if self.buffer.len() >= self.min_part_size as usize {
+                    self.upload_part().await?;
+                }
+            }
+        }
+        Ok(SinkReply::ACK)
+    }
+
+    // Required later
+    async fn on_signal(
+        &mut self,
+        _signal: Event,
+        _ctx: &SinkContext,
+        _serializer: &mut EventSerializer,
+    ) -> Result<SinkReply> {
+        // Check if a new client is required or not.
+        if let Ok(new_client) = self.client_receiver.try_recv() {
+            self.client = Some(new_client);
+        }
+        Ok(SinkReply::default())
+    }
+
+    async fn on_stop(&mut self, _ctx: &SinkContext) -> Result<()> {
+        // Commit the final upload.
+        if self.buffer.len() > 0 {
+            self.complete_multipart().await?;
+        }
+        Ok(())
+    }
+
+    fn asynchronous(&self) -> bool {
+        true
+    }
+
+    fn auto_ack(&self) -> bool {
+        false
+    }
+}
+
+#[async_trait::async_trait]
+impl Connector for S3Connector {
+    /// Currently no source
+    async fn create_source(
+        &mut self,
+        _source_context: SourceContext,
+        _builder: SourceManagerBuilder,
+    ) -> Result<Option<SourceAddr>> {
+        Ok(None)
+    }
+
+    /// Stream the events to the bucket
+    async fn create_sink(
+        &mut self,
+        sink_context: SinkContext,
+        builder: SinkManagerBuilder,
+    ) -> Result<Option<SinkAddr>> {
+        // FIXME: May be wise to increase channel size if you want to have multiple clients.
+        let (client_sender, client_receiver) = channel::bounded(2);
+
+        // set the sender and receiver for a new client
+        self.client_sender = Some(client_sender);
+
+        let s3_sink = S3Sink {
+            client_receiver,
+            bucket: self.config.bucket.clone(),
+            client: None,
+            buffer: Vec::new(),
```

Comment:
to avoid re-allocations and copying you could do a `Vec::with_capacity(self.config.min_part_size)`. It might not avoid all copying, but at least while you grow the buffer.